### ISSUE TITLE
Not upgrading anything

### ISSUE THREAD

**Issue by XNinety9** [mas-cli/mas#252]
Hi everyone.

Lately I found out that mas doesn't upgrades anything, although some upgrades exist in AppStore:
<img width="923" alt="Screenshot 2019-11-15 at 10 22 11" src="https://user-images.githubusercontent.com/4655239/68931905-d0071980-0791-11ea-9fdf-63f3560f5178.png">

My environment:

- macOS Catalina 1015.1
- mas 1.6.3, installed from brew

Thanks a lot for your help

**Comment by hexid26** [comment]
The same here, MacOS 10.15.1 (19B88).
`mas` installed from `brew`. (up-to-date)
`mas outdated` list nothing.
`mas upgrade` always display *Everything is up-to-date*.
However, `mas list`, `mas info` can get the correct info.
Lucky, `mas lucky` works fine.😎
<img width="742" alt="Screen Shot 2019-11-17 at 3 20 07 PM" src="https://user-images.githubusercontent.com/11417927/69004426-ebb71f00-094d-11ea-9b25-f886196ce5b6.png">

**Comment by gbstan** [comment]
Same issue on macox 10.15.2.
mas search, mas upgrade do not work any more

**Comment by nevyn-lookback** [comment]
Same on Mac OS 10.14.4 with mas 1.6.3. Tons of outdated software in my app store, mas upgrade says everything is up to date.

**Comment by garrettoreilly** [comment]
Ditto. macOS 10.15.4 and mas 1.6.3 from Homebrew. `mas outdated` returns nothing and `mas upgrade` returns `Everything is up-to-date`, even when there are updates available in the App Store.

**Comment by johnalanwoods** [comment]
same, totally broken.

**Comment by StarWitch** [comment]
Same issue here.

**Comment by waltmck** [comment]
Same exact issue here. Running Catalina 10.15.4 with mas 1.6.3, homebrew 2.2.14.

**Comment by captcarl13** [comment]
Still present in 1.6.4

**Comment by johnalanwoods** [comment]
Anyone else find it weird development is ongoing yet no one seems to notice this from dev?

**Comment by nevyn** [comment]
@johnalanwoods Not at all, this is OSS software and Ben has no obligation to us, and hard-to-reproduce bugs are a PITA.   The only thing that I find weird is that none of us has provided a PR. I’ll try to debug it next time I run into it...

**Comment by johnalanwoods** [comment]
@nevyn oh! I hadn't realised it was OSS! apologies!

It's not an esoteric bug. It affects 30 ppl in my office and all of my friends.

smfh.

**Comment by garrettoreilly** [comment]
@johnalanwoods @nevyn C’mon, we’re not children.

> Ben has no obligation to us, and hard-to-reproduce bugs are a PITA.

While true, this issue doesn’t seem difficult to reproduce.

**Comment by tpulatha** [comment]
It’s easy to reproduce but not easy to fix. There is a discussion here: https://github.com/mas-cli/mas/issues/248

This is OSS. Nobody owes you anything. Not even fixing easy bugs. If you are annoyed, you can help by going ahead and fixing the bug.

**Comment by johnalanwoods** [comment]
@tpulatha LOL

**Comment by phatblat** [comment]
The `outdated` and `upgrade` commands have been broken for quite some time. I believe they were flaky on Mojave. There is some discussion in #248 on a possible new macOS subsystem that may be responsible for this, but it is considerably more complicated to interface with.

**Comment by johnalanwoods** [comment]
@phatblat  - are they not the most useful functions though?

**Comment by tpulatha** [comment]
Indeed I agree that this app is pretty useless for my use case without being able to upgrade.

**Comment by Blaisorblade** [comment]
> It’s easy to reproduce but not easy to fix.

Easy improvement: add a print statement informing users they are known broken on some OS versions and pointing to the right issue.

That doesn't mean anybody _has_ to do it, it just means we should consider treat `mas` as unsupported and uninstall it.

**Comment by johnalanwoods** [comment]
I think this bug is a showstopper for the project. What use is a Mac App Store CLI that doesn't let you upgrade your apps.

**Comment by 454de6e** [comment]
> I think this bug is a showstopper for the project. What use is a Mac App Store CLI that doesn't let you upgrade your apps. 

The project is still super useful when you install a new laptop and script installation with something like Strap.

**Comment by johnalanwoods** [comment]
> > I think this bug is a showstopper for the project. What use is a Mac App Store CLI that doesn't let you upgrade your apps. 
> 
> The project is still super useful when you install a new laptop and script installation with something like Strap.

Ah, I see your point there.

**Comment by p-vernaeckt** [comment]
> > I think this bug is a showstopper for the project. What use is a Mac App Store CLI that doesn't let you upgrade your apps.
> 
> The project is still super useful when you install a new laptop and script installation with something like Strap.

If you like Strap, then you might want to checkout [reinstall.sh](https://gitlab.com/X99/reinstall.sh), a project of mine. It "uses" mas.

**Comment by phatblat** [comment]
Should be fixed now in [1.8.1](https://github.com/mas-cli/mas/releases/tag/v1.8.1). @p-vernaeckt can you confirm?

**Comment by p-vernaeckt** [comment]
It works here! Thanks!

**Comment by maiertech** [comment]
It also works for me, except that I'm seeing a false positive update notification for DaisyDisk:

```
411643860 DaisyDisk (4.12.1 -> 4.12)
```

I verified that installed version is 4.12.1, but AppStore shows 4.12.

**Comment by ZepDonald** [comment]
First of all, I'm glad, mas-cli is still under active development, hurray! 🎉

Second, I have exactly the same problem as @maiertech, but with more apps.

After entering `mas outdated` I get the following output:

```
Warning: Identifier 969978718 not found in store. Was expected to identify Noiseless.
1526844137 Search Key (1.8.0 -> 1.8)
1466810215 Murder Mystery Machine (1.17 -> 1.5)
1476088902 Rosetta Stone (7.14.0 -> 7.15.0)
1146204813 Comet (2.3.3 -> 2.3.2)
Warning: Identifier 716854513 not found in store. Was expected to identify Intensify.
1515959873 Theu2007Survivalists (1.0.684211 -> 1.2)
1489903517 lumen. (0.7 -> 1.3)
863486266 SketchBook (8.7.0 -> 8.7)
1286836110 WHITEWALL (1.3.6 -> 1.4.1)
1365531024 1Blocker (3.14 -> 3.14.1)
Warning: Identifier 446366603 not found in store. Was expected to identify ifolor Designer.
1083965373 Filters for Photos (1.0.0 -> 1.0)
1484348796 Endel (2.3.137 -> 2.3.2)
1136800740 GameTrack (3.2 -> 3.5)
Warning: Identifier 1438657064 not found in store. Was expected to identify NewsGuard.
1473627416 TheBradwellConspiracy (176.1384 -> 1.1)
1219074514 Vectornator (3.5.16 -> 3.5.15)
Warning: Identifier 872698314 not found in store. Was expected to identify MoneyMoney.
405399194 Kindle (1.30.0 -> 1.30)
Warning: Identifier 1063663640 not found in store. Was expected to identify Better Rename 10.
1495787023 ImageTextOCR (23.0 -> 23)
Warning: Identifier 480623975 not found in store. Was expected to identify Snapheal.
Warning: Identifier 892581529 not found in store. Was expected to identify Tonality.
1460715987 Oceanhorn 2 (100.78676.201122 -> 2.2)
```

I ran `mas upgrade` wich seems to work fine:

```
Upgrading 17 outdated applications:
Search Key (1.8.0), Murder Mystery Machine (1.17), Rosetta Stone (7.14.0), Comet (2.3.3), Theu2007Survivalists (1.0.684211), lumen. (0.7), SketchBook (8.7.0), WHITEWALL (1.3.6), 1Blocker (3.14), Filters for Photos (1.0.0), Endel (2.3.137), GameTrack (3.2), TheBradwellConspiracy (176.1384), Vectornator (3.5.16), Kindle (1.30.0), ImageTextOCR (23.0), Oceanhorn 2 (100.78676.201122)
==> Downloading Search Key
==> Installed Search Key
==> Downloading Murder Mystery Machine
==> Installed Murder Mystery Machine
==> Downloading Rosetta Stone: Sprachen lernen
==> Installed Rosetta Stone: Sprachen lernen
==> Downloading Comet for reddit
==> Installed Comet for reddit
==> Downloading The Survivalists
==> Installed The Survivalists
==> Downloading lumen.
==> Installed lumen.
==> Downloading Autodesk SketchBook
==> Installed Autodesk SketchBook
==> Downloading WhiteWall für Ihre Fotos
==> Installed WhiteWall für Ihre Fotos
==> Downloading 1Blocker
==> Installed 1Blocker
==> Downloading Filters for Photos
==> Installed Filters for Photos
==> Downloading Endel: Focus, Sleep, Relax
==> Installed Endel: Focus, Sleep, Relax
==> Downloading GameTrack
==> Installed GameTrack
==> Downloading The Bradwell Conspiracy
==> Installed The Bradwell Conspiracy
==> Downloading Vectornator - Grafikdesign
==> Installed Vectornator - Grafikdesign
==> Downloading Kindle
==> Installed Kindle
==> Downloading Image Text OCR Photo, PDF Scan
==> Installed Image Text OCR Photo, PDF Scan
==> Downloading Oceanhorn 2
==> Installed Oceanhorn 2
```

But `mas outdated` still gives me the aforementioned output.

And some apps can't even be found. While for some there is no Store page anymore, I do have at least one app, where this is not the case:

`Warning: Identifier 872698314 not found in store. Was expected to identify MoneyMoney.`

Btw. `mas version` is `1.8.1`, just to be clear.

I hope this is easy to be fixed. ☺️

**Comment by rgoldberg** [comment]
I implemented both commands to consider any version mismatch to be outdated. If some versions have version numbers with more than digits & dots, then version ordering can get tricky. e.g., Maven's version ordering is buggy, non-exhaustive, and poorly orders some weird versions.

@ZepDonald It's odd that mas reports that the store versions of so many of your apps are at versions that are older than what you've installed. Can you find out if the store versions & the local versions reported by mas are correct?

**Comment by maiertech** [comment]
This might be some bug with the AppStore. For DaisyDisk, the AppStore shows version 4.12:

<img width="929" alt="Screen Shot 2021-03-15 at 5 37 23 PM" src="https://user-images.githubusercontent.com/1482402/111224794-4cdc1a80-85b5-11eb-9503-f148e512051c.png">

`mas search DaisyDisk` returns

```
411643860  DaisyDisk (4.12)
```

When I remove DaisyDisk and install with `mas` or directly from the AppStore, I get version 4.12.1:

<img width="248" alt="Screen Shot 2021-03-15 at 5 42 10 PM" src="https://user-images.githubusercontent.com/1482402/111225173-d855ab80-85b5-11eb-8c0f-5b8fc1fee7a7.png">

And then `mas outdated` says

```
411643860 DaisyDisk (4.12.1 -> 4.12)
```

and the AppStore does not show that there is an update, but keeps displaying 4.12.

**Comment by ZepDonald** [comment]
Sorry for the late response.

@rgoldberg I will try my best! 😊

For the wrong versions, I tried with just a few of the apps, as all results are pretty much the same and I can confirm the statement from @maiertech.

**E.g. Search Key**

```
$ mas outdated
1526844137 Search Key (1.8.0 -> 1.8)
$ mas search 1526844137
1526844137  Search Key (1.8)
```

According to the App Store, the version is also 1.8.
In Finder, the App has version 1.8.0.

**Another example: Endel**

```
$ mas outdated
1484348796 Endel (2.3.137 -> 2.3.2)
$ mas search 1484348796
1484348796  Endel: Focus, Sleep, Relax (2.3.2)
```

App Store: 2.3.2
Finder: 2.3.137

For MoneyMoney, `mas search MoneyMoney` shows me around 45 Apps, but not MoneyMoney.
I also tried `mas search MoneyMoney | grep MoneyMoney` to be sure, I'm not overlooking something but with no results.
I copied the Link from the App Store Page, which shows the correct ID:
https://apps.apple.com/de/app/moneymoney/id872698314?mt=12

For comparison, that's the relevant output of `mas outdated`:
```
Warning: Identifier 872698314 not found in store. Was expected to identify MoneyMoney.
```

I'm located at Germany and my store is set to the German store.

If there's more input needed, just let me know.

**Comment by chris-araman** [comment]
The issues around discrepancies between `outdated` and `upgrade` should be resolved by #346.

The issue with MoneyMoney has to do with the app only being available in Germany. We don't currently pass `country` to the App Store search API. That work is tracked by #75.
